### PR TITLE
added bucket sort algorithm

### PR DESCRIPTION
**Pull request by DrakonDarquesse**
Hi sir/madam,

I am Ming Liang and I have implemented a sorting algorithm called Bucket sort.
I hope you can take a moment to read the algorithm. Thank you in advance.

Best regards,
Ming Liang

### CODE REVIEW

File: src/sorts/bucket_sort.jl [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+                idx = idx - 1
+            end
+
+            bucket[idx] = val 
+        end
+    end
+    arr = vcat(buckets...)
+end
+
+BucketSort!([1,2,3])
```
Comment (Panquesito7) [inline src/sorts/bucket_sort.jl:68]:
```suggestion
BucketSort!([1,2,3])

```

File: src/sorts/bucket_sort.jl [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+number of buckets = length of array
+size of buckets = (largest element - smallest element + 1) / length of array
+
+## Bucket Sort Steps
+1. Create an array of buckets
+2. Add each elements into its respective bucket
+3. Sort each non-empty bucket
+4. Concatenate elements of non-empty buckets
+
+## Example
```
Comment (Panquesito7) [inline src/sorts/bucket_sort.jl:19]:
```suggestion
## Example

```

File: src/sorts/bucket_sort.jl [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+Bucket Sort is a divide and conquer sorting algorithm.
+It distributes elements into buckets and sort each bucket before merging all buckets
+Buckets are arrays that each stores a different range of elements
+Insertion sort is used to sort the buckets because its fast in small-sized arrays
+
+## Bucket Calculation
+number of buckets = length of array
+size of buckets = (largest element - smallest element + 1) / length of array
+
+## Bucket Sort Steps
```
Comment (Panquesito7) [inline src/sorts/bucket_sort.jl:13]:
```suggestion
## Bucket Sort Steps

```

File: src/sorts/bucket_sort.jl [hunk 4]
```diff
@@ -0,0 +1,68 @@
+"""
+# Bucket Sort
+
+Bucket Sort is a divide and conquer sorting algorithm.
+It distributes elements into buckets and sort each bucket before merging all buckets
+Buckets are arrays that each stores a different range of elements
+Insertion sort is used to sort the buckets because its fast in small-sized arrays
+
+## Bucket Calculation
```
Comment (Panquesito7) [inline src/sorts/bucket_sort.jl:9]:
```suggestion
## Bucket Calculation

```

**Review by Panquesito7 — APPROVED** [review]
Thanks.

**Comment by Azzaare** [comment]
@DrakonDarquesse to make the tests possible, you have to add `BucketSort!` to the export list in the TheAlgorithms.jl file.

**Comment by uncomfyhalomacro** [comment]
@DrakonDarquesse Hello. Sorry for the ping. What is the progress of this PR?

**Comment by DrakonDarquesse** [comment]
Will complete it asap

**Comment by Azzaare** [comment]
@DrakonDarquesse I have finished to fix the PR if you want

**Comment by Azzaare** [comment]
If you merge https://github.com/DrakonDarquesse/Julia/pull/1 it should fix this PR

**Review by Panquesito7 — COMMENTED** [review]
Any reasons why the CodeCoverage is failing? 🤔

image

**Comment by Azzaare** [comment]
> Any reasons why the CodeCoverage is failing? thinking
> 
> image

Not all the code introduced was tested.  A couple of line were not hit by the tests, thus not covered.

I think it is okay in this case. On the long run, it is likely that we should implement extensive testing for sorting algorithms.

**Review by Azzaare — APPROVED** [review]

**Pull request merged**